src/db_reader.rs

```diff
@@ -0,0 +1,1083 @@
+use crate::bytes_range::BytesRange;
+use crate::config::{
+    CheckpointOptions, Clock, DbReaderOptions, ReadOptions, ScanOptions, SystemClock,
+};
+use crate::db_reader::ManifestPollerMsg::Shutdown;
+use crate::db_state::{COWDbState, CoreDbState, DbStateSnapshot};
+use crate::db_stats::DbStats;
+use crate::error::SlateDBError;
+use crate::manifest::store::{DirtyManifest, ManifestStore, StoredManifest};
+use crate::mem_table::{ImmutableMemtable, KVTable};
+use crate::reader::{Reader, ReaderStateSupplier};
+use crate::sst::SsTableFormat;
+use crate::sst_iter::SstIteratorOptions;
+use crate::stats::StatRegistry;
+use crate::tablestore::TableStore;
+use crate::utils::MonotonicClock;
+use crate::wal_replay::{WalReplayIterator, WalReplayOptions};
+use crate::{utils, Checkpoint, DbIterator};
+use bytes::Bytes;
+use log::{info, warn};
+use object_store::path::Path;
+use object_store::ObjectStore;
+use parking_lot::{Mutex, RwLock};
+use std::collections::VecDeque;
+use std::ops::{RangeBounds, Sub};
+use std::sync::Arc;
+use tokio::runtime::Handle;
+use tokio::select;
+use tokio::sync::mpsc::{UnboundedReceiver, UnboundedSender};
+use uuid::Uuid;
+
+/// Read-only interface for accessing a database from either
+/// the latest persistent state or from an arbitrary checkpoint.
+pub struct DbReader {
+    inner: Arc<DbReaderInner>,
+    manifest_poller: Option<ManifestPoller>,
+}
+
+struct DbReaderInner {
+    manifest_store: Arc<ManifestStore>,
+    table_store: Arc<TableStore>,
+    options: DbReaderOptions,
+    state: Arc<RwLock<CheckpointState>>,
+    clock: Arc<dyn Clock + Sync + Send>,
+    reader: Reader,
+}
+
+impl ReaderStateSupplier for RwLock<CheckpointState> {
+    fn supply(&self) -> DbStateSnapshot {
+        let guard = self.read();
+        let state = guard.state.clone();
+        DbStateSnapshot {
+            memtable: Arc::new(KVTable::new()),
+            wal: Arc::new(KVTable::new()),
+            state,
+        }
+    }
+}
+
+struct ManifestPoller {
+    join_handle: Mutex<Option<tokio::task::JoinHandle<Result<(), SlateDBError>>>>,
+    thread_tx: UnboundedSender<ManifestPollerMsg>,
+}
+
+enum ManifestPollerMsg {
+    Shutdown,
+}
+
+#[derive(Clone)]
+struct CheckpointState {
+    checkpoint: Checkpoint,
+    state: Arc<COWDbState>,
```

Comment:
Moving our offline discussion to the PR. You raised that it's awkward to hold a `DirtyManifest` in the reader's `CheckpointState`. One way to avoid this would be to decouple `CheckpointState` from `COWDbState` by adding a trait like:
```rust
trait ReadSnapshot {
    fn memtable(&self) -> Arc<KVTable>,
    fn wal(&self) -> Arc<KVTable>,
    fn imm_memtable(&self) -> VecDequeue<Arc<ImmutableMemtable>>,
    fn imm_wal(&self) -> VecDequeue<Arc<ImmutableWal>>,
    fn core(&self) -> &CoreDbState
}
```

The writer and reader can each have their own impl for that, and the reader's impl can just hold a `Manifest` directly.

Reply:
I think I tried something like this. The challenge is working with `Arc<COWDbState>`. I don't think it's possible to get a `VecDeque<Arc<ImmutableMemtable>>` (for example) out of this without cloning the `VecDeque` in `COWDbState`. This felt a little too costly for a `get`. Perhaps it would work if we could flatten `COWDbState` in `DbState`:

```
pub(crate) struct DbState {
    memtable: WritableKVTable,
    wal: WritableKVTable,
    imm_memtable: Arc<VecDeque<Arc<ImmutableMemtable>>>,
    imm_wal: Arc<VecDeque<Arc<ImmutableWal>>>,
    manifest: Arc<DirtyManifest>,    
    last_seq: u64,
    error: WatchableOnceCell<SlateDBError>,
}
```
Then I think the trait might work out. But maybe there's something simpler that I'm missing?

Reply:
On second consideration, I think this might work if we just return references to the VecDeques